L1Trigger/L1TZDC/plugins/L1TZDCProducer.cc

```diff
@@ -0,0 +1,4885 @@
+
+#include "DataFormats/HcalDigi/interface/HcalDigiCollections.h"
+
+//
+// class declaration
+//
+
+using namespace l1t;
+
+class L1TZDCProducer : public edm::stream::EDProducer<> {
+public:
+  explicit L1TZDCProducer(const edm::ParameterSet& ps);
+  ~L1TZDCProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  void beginRun(edm::Run const&, edm::EventSetup const&) override;
+  void endRun(edm::Run const&, edm::EventSetup const&) override;
+
+  int zdcLUTIndexHelper(int iDetPos, int iBXPos);
+
+  //virtual void beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+  //virtual void endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+
+  // ----------member data ---------------------------
+
+  // input tokens
+  //Add the ZDC token - remove other tokens later when more certain what is not needed i.e. at direction of L1Trigger maintainers, particularly the Stage2Layer2Producer maintainers
+  edm::EDGetTokenT<QIE10DigiCollection> m_zdcToken;
+  edm::ESGetToken<CaloParams, L1TCaloParamsRcd> m_candidateToken;
+
+  // put tokens
+  edm::EDPutTokenT<EtSumBxCollection> m_etToken;
+
+  //Following the L1TStage2Layer2Producer
+  CaloParamsHelper* m_params;
+  int m_scaleFactor;
+
+  bool m_doHardCodeLUT;
+};
+
+L1TZDCProducer::L1TZDCProducer(const edm::ParameterSet& ps) {
+  // register what you produce
+  m_etToken = produces<EtSumBxCollection>("zdcEtSums");
+
+  // register what you consume and keep token for later access:
+  m_zdcToken = consumes<QIE10DigiCollection>(ps.getParameter<edm::InputTag>("zdcToken"));
+  m_candidateToken = esConsumes<CaloParams, L1TCaloParamsRcd, edm::Transition::BeginRun>();
+
+  // placeholder for the parameters
+  m_params = new CaloParamsHelper;
+
+  m_doHardCodeLUT = ps.getParameter<bool>("doHardCodeLUT");
+}
+
+L1TZDCProducer::~L1TZDCProducer() { delete m_params; }
+
+// ------------ method called to produce the data  ------------
+void L1TZDCProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  //std::cout<<"-------------------------------- New event ---------------------"<<std::endl;
+  using namespace edm;
+
+  using namespace l1t;
+
+  LogDebug("l1t|stage 2") << "L1TZDCProducer::produce function called..." << std::endl;
+
+  //Below is hardcoded LUT for testing purposes
+  //A python based LUT also exists under L1Trigger/L1TZDC/data
+  //18 indices: NEM1, NEM2, NEM3, NEM4, NEM5, NHD1, NHD2, NHD3, NHD4, PEM1, PEM2, PEM3, PEM4, PEM5, PHD1, PHD2, PHD3, PHD4
+  double QIE10_regular_fC_full[256][18] = {{0.16599,
+                                            0.373632,
+                                            0.381661,
+                                            0.31559,
+                                            0.322258,
+                                            1.62607,
```

Comment:
This "m_doHardCodeLUT" option leading to a plugin with 4000+ lines of hard-coded numbers does not look like something that should be integrated centrally. Plugins do not normally contain "magic numbers" unless well-justified (certainly not thousands of them). These numbers, if needed, should likely be moved elsewhere (cms-data ? conditions database ? else ?).

Reply:
@missirol happy to remove; leftover from a testing stage, wasn't sure what to do with it. 

Since you are also commenting on the paired PR to this one by @elfontan (see here: https://github.com/cms-sw/cmssw/pull/42635) perhaps you could comment on our proposal to resolve issues in that branch here: https://github.com/cms-sw/cmssw/pull/42634/#issuecomment-1691143781

Does it seem like a correct way to resolve?

Reply:
Short answer: if it was me, I would probably update this PR including #42635.

Longer answer:

> The PR parallel to this one for the GT (found here: https://github.com/cms-sw/cmssw/pull/42635 ) is running into issues in build tests because it has dependencies on this branch that weren't incorporated

That's not my understanding. The latest error in #42635 looks like a simple technicality [(an unused variable to be removed)](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4995a3/34457/clang-new-warnings.log). At the same time, the two PRs do overlap. My understanding is that both need the changes in `DataFormats`, and I don't see an easy way to disentangle them. You could have a first PR (incl. data-format changes) integrated first, then followed by the 2nd PR. On the other hand, time looks like an important factor in this case, and the two PRs look closely related and both necessary (and all of this needs to be backported). Given all the above, I would keep this PR open (at least we don't lose the review that started here), update it with the content of #42635, and close the latter.

If you keep this PR open, it would be useful to improve its title.

Reply:
This is removed in latest commits